do not use deepcopier to create a deep copy of gRPC structs

Looks like deepcopier is not able to create a deep copy of a nested pointer grpc type
```
func TestDeepcopier(t *testing.T) {
	t.Run("test ptr in deepcopier", func(t *testing.T) {
		f1 := &pb.EncryptedVolume{VolumeId: &pc.ObjectKey{Value: "abc"}}
		f2 := &pb.EncryptedVolume{}
		deepcopier.Copy(f1).To(f2)
		log.Println(f1)
		log.Println(f2)
		log.Printf("%p %p", f1.VolumeId, f2.VolumeId)
		if !proto.Equal(f1, f2) {
			t.Errorf("Expect deep copy of structure %v, got %v", f1, f2)
		}
		if f1.VolumeId == f2.VolumeId {
			t.Errorf("f1 vol ptr cannot be equal to f2 vol ptr")
		}
	})
}
```
shows
```
2023/05/29 11:54:05 volume_id:{value:"abc"}
2023/05/29 11:54:05 volume_id:{value:"abc"}
2023/05/29 11:54:05 0xc0000c9c00 0xc0000c9c00
--- FAIL: TestDeepcopier (0.00s)
    --- FAIL: TestDeepcopier/test_ptr_in_deepcopier (0.00s)
        /home/abc/test/test_test.go:518: f1 vol ptr cannot be equal to f2 vol ptr
FAIL
```

For gRPC structures, [proto.Clone](https://pkg.go.dev/github.com/golang/protobuf/proto#Clone) should be used instead

--- a/go.mod
+++ b/go.mod
@@ -7,7 +7,6 @@ require (
 	github.com/google/uuid v1.3.0
 	github.com/opiproject/gospdk v0.0.0-20230515120524-37c85998ff39
 	github.com/opiproject/opi-api v0.0.0-20230525195117-1db0f9cbae90
-	github.com/ulule/deepcopier v0.0.0-20200430083143-45decc6639b6
 	google.golang.org/grpc v1.55.0
 	google.golang.org/protobuf v1.30.0
 )

--- a/pkg/backend/aio.go
+++ b/pkg/backend/aio.go
@@ -16,7 +16,6 @@ import (
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
 
 	"github.com/google/uuid"
-	"github.com/ulule/deepcopier"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
@@ -62,12 +61,7 @@ func (s *Server) CreateAioController(_ context.Context, in *pb.CreateAioControll
 		log.Print(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
-	response := &pb.AioController{}
-	err = deepcopier.Copy(in.AioController).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.AioController)
 	s.Volumes.AioVolumes[in.AioController.Name] = response
 	return response, nil
 }
@@ -138,12 +132,7 @@ func (s *Server) UpdateAioController(_ context.Context, in *pb.UpdateAioControll
 		log.Print(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
-	response := &pb.AioController{}
-	err3 := deepcopier.Copy(in.AioController).To(response)
-	if err3 != nil {
-		log.Printf("error: %v", err3)
-		return nil, err3
-	}
+	response := server.ProtoClone(in.AioController)
 	s.Volumes.AioVolumes[in.AioController.Name] = response
 	return response, nil
 }

--- a/pkg/backend/null.go
+++ b/pkg/backend/null.go
@@ -17,7 +17,6 @@ import (
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
 
 	"github.com/google/uuid"
-	"github.com/ulule/deepcopier"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
@@ -63,12 +62,7 @@ func (s *Server) CreateNullDebug(_ context.Context, in *pb.CreateNullDebugReques
 		log.Print(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
-	response := &pb.NullDebug{}
-	err = deepcopier.Copy(in.NullDebug).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NullDebug)
 	s.Volumes.NullVolumes[in.NullDebug.Name] = response
 	return response, nil
 }
@@ -139,12 +133,7 @@ func (s *Server) UpdateNullDebug(_ context.Context, in *pb.UpdateNullDebugReques
 		log.Print(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
-	response := &pb.NullDebug{}
-	err3 := deepcopier.Copy(in.NullDebug).To(response)
-	if err3 != nil {
-		log.Printf("error: %v", err3)
-		return nil, err3
-	}
+	response := server.ProtoClone(in.NullDebug)
 	s.Volumes.NullVolumes[in.NullDebug.Name] = response
 	return response, nil
 }

--- a/pkg/backend/nvme.go
+++ b/pkg/backend/nvme.go
@@ -17,7 +17,6 @@ import (
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
 
 	"github.com/google/uuid"
-	"github.com/ulule/deepcopier"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
@@ -67,12 +66,7 @@ func (s *Server) CreateNVMfRemoteController(_ context.Context, in *pb.CreateNVMf
 	if len(result) != 1 {
 		log.Printf("expecting exactly 1 result")
 	}
-	response := &pb.NVMfRemoteController{}
-	err = deepcopier.Copy(in.NvMfRemoteController).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NvMfRemoteController)
 	s.Volumes.NvmeVolumes[in.NvMfRemoteController.Name] = response
 	return response, nil
 }

--- a/pkg/frontend/blk.go
+++ b/pkg/frontend/blk.go
@@ -16,7 +16,6 @@ import (
 	pc "github.com/opiproject/opi-api/common/v1/gen/go"
 	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
-	"github.com/ulule/deepcopier"
 
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
@@ -63,12 +62,7 @@ func (s *Server) CreateVirtioBlk(_ context.Context, in *pb.CreateVirtioBlkReques
 	}
 	s.Virt.BlkCtrls[in.VirtioBlk.Name] = in.VirtioBlk
 	// s.VirtioCtrls[in.VirtioBlk.Name].Status = &pb.NvmeControllerStatus{Active: true}
-	response := &pb.VirtioBlk{}
-	err = deepcopier.Copy(in.VirtioBlk).To(response)
-	if err != nil {
-		log.Printf("Error at response creation: %v", err)
-		return nil, status.Error(codes.Internal, "Failed to construct device create response")
-	}
+	response := server.ProtoClone(in.VirtioBlk)
 	return response, nil
 }
 

--- a/pkg/frontend/nvme.go
+++ b/pkg/frontend/nvme.go
@@ -17,7 +17,6 @@ import (
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
 
 	"github.com/google/uuid"
-	"github.com/ulule/deepcopier"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
@@ -136,12 +135,7 @@ func (s *Server) CreateNvmeSubsystem(_ context.Context, in *pb.CreateNvmeSubsyst
 		return nil, err
 	}
 	log.Printf("Received from SPDK: %v", ver)
-	response := &pb.NvmeSubsystem{}
-	err = deepcopier.Copy(in.NvmeSubsystem).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NvmeSubsystem)
 	response.Status = &pb.NvmeSubsystemStatus{FirmwareRevision: ver.Version}
 	s.Nvme.Subsystems[in.NvmeSubsystem.Spec.Name] = response
 	return response, nil
@@ -297,12 +291,8 @@ func (s *Server) CreateNvmeController(_ context.Context, in *pb.CreateNvmeContro
 	s.Nvme.Controllers[in.NvmeController.Spec.Name] = in.NvmeController
 	s.Nvme.Controllers[in.NvmeController.Spec.Name].Spec.NvmeControllerId = -1
 	s.Nvme.Controllers[in.NvmeController.Spec.Name].Status = &pb.NvmeControllerStatus{Active: true}
-	response := &pb.NvmeController{Spec: &pb.NvmeControllerSpec{Name: "TBD"}}
-	err = deepcopier.Copy(in.NvmeController).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NvmeController)
+
 	return response, nil
 }
 
@@ -347,12 +337,7 @@ func (s *Server) UpdateNvmeController(_ context.Context, in *pb.UpdateNvmeContro
 	log.Printf("UpdateNvmeController: Received from client: %v", in)
 	s.Nvme.Controllers[in.NvmeController.Spec.Name] = in.NvmeController
 	s.Nvme.Controllers[in.NvmeController.Spec.Name].Status = &pb.NvmeControllerStatus{Active: true}
-	response := &pb.NvmeController{}
-	err := deepcopier.Copy(in.NvmeController).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NvmeController)
 	return response, nil
 }
 
@@ -433,12 +418,7 @@ func (s *Server) CreateNvmeNamespace(_ context.Context, in *pb.CreateNvmeNamespa
 	}
 	s.Nvme.Namespaces[in.NvmeNamespace.Spec.Name] = in.NvmeNamespace
 
-	response := &pb.NvmeNamespace{}
-	err = deepcopier.Copy(in.NvmeNamespace).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NvmeNamespace)
 	response.Status = &pb.NvmeNamespaceStatus{PciState: 2, PciOperState: 1}
 	return response, nil
 }
@@ -488,12 +468,7 @@ func (s *Server) UpdateNvmeNamespace(_ context.Context, in *pb.UpdateNvmeNamespa
 	s.Nvme.Namespaces[in.NvmeNamespace.Spec.Name] = in.NvmeNamespace
 	s.Nvme.Namespaces[in.NvmeNamespace.Spec.Name].Status = &pb.NvmeNamespaceStatus{PciState: 2, PciOperState: 1}
 
-	response := &pb.NvmeNamespace{}
-	err := deepcopier.Copy(in.NvmeNamespace).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.NvmeNamespace)
 	return response, nil
 }
 

--- a/pkg/frontend/scsi.go
+++ b/pkg/frontend/scsi.go
@@ -17,7 +17,6 @@ import (
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
 
 	"github.com/google/uuid"
-	"github.com/ulule/deepcopier"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
@@ -61,12 +60,7 @@ func (s *Server) CreateVirtioScsiController(_ context.Context, in *pb.CreateVirt
 	}
 	s.Virt.ScsiCtrls[in.VirtioScsiController.Name] = in.VirtioScsiController
 	// s.VirtioCtrls[in.VirtioScsiController.Name].Status = &pb.VirtioScsiControllerStatus{Active: true}
-	response := &pb.VirtioScsiController{}
-	err = deepcopier.Copy(in.VirtioScsiController).To(response)
-	if err != nil {
-		log.Printf("Error at response creation: %v", err)
-		return nil, status.Error(codes.Internal, "Failed to construct device create response")
-	}
+	response := server.ProtoClone(in.VirtioScsiController)
 	return response, nil
 }
 
@@ -198,12 +192,7 @@ func (s *Server) CreateVirtioScsiLun(_ context.Context, in *pb.CreateVirtioScsiL
 	log.Printf("Received from SPDK: %v", result)
 	s.Virt.ScsiLuns[in.VirtioScsiLun.Name] = in.VirtioScsiLun
 	// s.ScsiLuns[in.VirtioScsiLun.Name].Status = &pb.VirtioScsiLunStatus{Active: true}
-	response := &pb.VirtioScsiLun{}
-	err = deepcopier.Copy(in.VirtioScsiLun).To(response)
-	if err != nil {
-		log.Printf("Error at response creation: %v", err)
-		return nil, status.Error(codes.Internal, "Failed to construct device create response")
-	}
+	response := server.ProtoClone(in.VirtioScsiLun)
 	return response, nil
 }
 

--- a/pkg/kvm/blk_test.go
+++ b/pkg/kvm/blk_test.go
@@ -8,14 +8,13 @@ import (
 	"bytes"
 	"context"
 	"errors"
-	"log"
 	"testing"
 
 	"github.com/opiproject/gospdk/spdk"
 	pc "github.com/opiproject/opi-api/common/v1/gen/go"
 	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
 	"github.com/opiproject/opi-spdk-bridge/pkg/frontend"
-	"github.com/ulule/deepcopier"
+	"github.com/opiproject/opi-spdk-bridge/pkg/server"
 	"google.golang.org/protobuf/proto"
 )
 
@@ -31,10 +30,7 @@ var (
 )
 
 func TestCreateVirtioBlk(t *testing.T) {
-	expectNotNilOut := &pb.VirtioBlk{}
-	if deepcopier.Copy(testCreateVirtioBlkRequest.VirtioBlk).To(expectNotNilOut) != nil {
-		log.Panicf("Failed to copy structure")
-	}
+	expectNotNilOut := server.ProtoClone(testCreateVirtioBlkRequest.VirtioBlk)
 	expectNotNilOut.Name = testVirtioBlkID
 
 	tests := map[string]struct {

--- a/pkg/middleend/encryption.go
+++ b/pkg/middleend/encryption.go
@@ -16,7 +16,6 @@ import (
 	"github.com/opiproject/gospdk/spdk"
 	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
 	"github.com/opiproject/opi-spdk-bridge/pkg/server"
-	"github.com/ulule/deepcopier"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/emptypb"
@@ -76,12 +75,7 @@ func (s *Server) CreateEncryptedVolume(_ context.Context, in *pb.CreateEncrypted
 		log.Print(msg)
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
-	response := &pb.EncryptedVolume{}
-	err = deepcopier.Copy(in.EncryptedVolume).To(response)
-	if err != nil {
-		log.Printf("error: %v", err)
-		return nil, err
-	}
+	response := server.ProtoClone(in.EncryptedVolume)
 	return response, nil
 }
 
@@ -194,12 +188,7 @@ func (s *Server) UpdateEncryptedVolume(_ context.Context, in *pb.UpdateEncrypted
 		return nil, status.Errorf(codes.InvalidArgument, msg)
 	}
 	// return result
-	response := &pb.EncryptedVolume{}
-	err4 := deepcopier.Copy(in.EncryptedVolume).To(response)
-	if err4 != nil {
-		log.Printf("error: %v", err4)
-		return nil, err4
-	}
+	response := server.ProtoClone(in.EncryptedVolume)
 	return response, nil
 }
 

--- a/pkg/server/utils.go
+++ b/pkg/server/utils.go
@@ -18,6 +18,7 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
 
 	"github.com/opiproject/gospdk/spdk"
 )
@@ -139,4 +140,9 @@ func spdkMockServerCommunicate(rpc spdk.JSONRPC, l net.Listener, toSend []string
 			log.Fatal(err)
 		}
 	}
+}
+
+// ProtoClone creates a deep copy of a provided gRPC structure
+func ProtoClone[T proto.Message](protoStruct T) T {
+	return proto.Clone(protoStruct).(T)
 }

--- a/pkg/server/utils_test.go
+++ b/pkg/server/utils_test.go
@@ -1,0 +1,41 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Intel Corporation
+
+// Package server implements the server
+package server
+
+import (
+	"testing"
+
+	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
+	"google.golang.org/protobuf/proto"
+)
+
+func TestProtoClone(t *testing.T) {
+	tests := map[string]struct {
+		in *pb.NvmeController
+	}{
+		"proto structure": {
+			in: &pb.NvmeController{
+				Spec: &pb.NvmeControllerSpec{
+					PcieId: &pb.PciEndpoint{},
+				},
+			},
+		},
+		"nil proto structure": {
+			in: nil,
+		},
+	}
+	for testName, tt := range tests {
+		t.Run(testName, func(t *testing.T) {
+			copiedIn := ProtoClone(tt.in)
+
+			if !proto.Equal(tt.in, copiedIn) {
+				t.Errorf("Expect proto structure copy %v, received: %v", tt.in, copiedIn)
+			}
+			if tt.in != nil && tt.in.Spec.PcieId == copiedIn.Spec.PcieId {
+				t.Errorf("Expect deep copy, not pointer copy")
+			}
+		})
+	}
+}

